001-spin-kube-doctor.md

```diff
@@ -0,0 +1,218 @@
+
+
+```
+type CheckFn func(ctx context.Context, k Provider, check Check) (Status, error)
+```
+
+An example of how to check if a crd is installed:
+
+```
+var isCrdInstalled = func(ctx context.Context, k provider.Provider, check provider.Check) (provider.Status, error) {
+        _, err := k.DynamicClient().Resource(schema.GroupVersionResource{
+                Group:    "apiextensions.k8s.io",
+                Version:  "v1",
+                Resource: "customresourcedefinitions",
+        }).Get(ctx, check.ResourceName, metav1.GetOptions{})
+        if err != nil {
+                if errors.IsNotFound(err) {
+                        return provider.Status{
+                                Name:     check.Name,
+                                Ok:       false,
+                        }, nil
+                }
+
+                return provider.Status{
+                        Name:     check.Name,
+                        Ok:       false,
+                        HowToFix: check.HowToFix,
+                }, err
+        }
+
+        return provider.Status{
+                Name: check.Name,
+                Ok:   true,
+        }, nil
+}
+
+```
+
+We can then have a default set of checks that can be performed on the cluster with the possibility of customizing them if required. 
+
+### how multiple providers are supported
+
+Because these checks will be using Kubernetes API to perform checks, they should work on most of the Kubernetes distribution providers by default. However, in case we need to override the checks for a given distribution, we can do so by implementing the following interface for that provider (specifically the GetCheckOverride function):
+
+```
+type Provider interface {
+        Name() string
+        Client() kubernetes.Interface
+        DynamicClient() dynamic.Interface
+        Status(ctx context.Context) ([]Status, error)
+        GetCheckOverride(ctx context.Context, check Check) CheckFn
+}
+
+```
+
+An example of this to adjust the check to support k3d is shown below:
+
+```
+func (k *k3d) GetCheckOverride(ctx context.Context, check provider.Check) provider.CheckFn {
+        switch check.Type {
+        case checks.CheckBinaryInstalledOnNodes:
+                return binaryVersionCheck
+        }
+
+        return nil
+}
+
+var binaryVersionCheck = func(ctx context.Context, k provider.Provider, check provider.Check) (provider.Status, error) {
+        return checks.ExecOnEachNodeFn(ctx, k, check, []string{"/host/bin/containerd-shim-spin-v2"}, []string{"-v"})
+}
+
+```
+
+#### Examples
+
+Here is how the output currently looks like for a few providers:
+
+##### k3d
+```
+$ go run main.go doctor --context k3d-wasm-cluster
+
+
+#-------------------------------------
+# Running checks for SpinKube setup
+#-------------------------------------
+
+✓ Containerd version is supported
+✓ Containerd shim is installed and configured
```

Comment:
Should we also check if the spin app itself is compatible with the shim? Say trigger type isn't supported.

Reply:
that is possible but, at this moment, these checks are not specific to any particular spin app. but seems like something that can be useful.